File: MiniDOMSIP/Config/Servers/AsteriskServer.cs
using DMD;
using DMD.XML;
using minidom.CallManagers;
using System;
using System.Runtime.CompilerServices;

namespace minidom.PBX
{

    /// <summary>
    /// Rappresenta un server asterisk
    /// </summary>
    [Serializable]
    public class AsteriskServer 
        : IDMDXMLSerializable
    {

        void IDMDXMLSerializable.SetFieldInternal(string fieldName, object fieldValue) { this.SetFieldInternal(fieldName, fieldValue);  }
        void IDMDXMLSerializable.XMLSerialize(XMLWriter writer) { this.XMLSerialize(writer); }

        public event ConnectedEventHandler Connected;

        public delegate void ConnectedEventHandler(object sender, AsteriskEventArgs e);

        public event DisconnectedEventHandler Disconnected;

        public delegate void DisconnectedEventHandler(object sender, AsteriskEventArgs e);

        public event ManagerEventEventHandler ManagerEvent;

        public delegate void ManagerEventEventHandler(object sender, AsteriskEvent e);

        private System.Timers.Timer _m_Timer;

        private System.Timers.Timer m_Timer
        {
            [MethodImpl(MethodImplOptions.Synchronized)]
            get
            {
                return _m_Timer;
            }

            [MethodImpl(MethodImplOptions.Synchronized)]
            set
            {
                if (_m_Timer != null)
                {
                    _m_Timer.Elapsed -= m_Timer_Elapsed;
                }

                _m_Timer = value;
                if (_m_Timer != null)
                {
                    _m_Timer.Elapsed += m_Timer_Elapsed;
                }
            }
        }

        public string Nome;
        public string ServerName;
        public int ServerPort = 5038;
        public string UserName = "";
        public string Password = "";
        public string Channel = "";
        public string CallerID = "";
        [NonSerialized] private AsteriskCallManager _m_Manager;

        private AsteriskCallManager m_Manager
        {
            [MethodImpl(MethodImplOptions.Synchronized)]
            get
            {
                return _m_Manager;
            }

            [MethodImpl(MethodImplOptions.Synchronized)]
            set
            {
                if (_m_Manager != null)
                {
                    _m_Manager.ManagerEvent -= handleAsteriskEvent;
                    _m_Manager.Connected -= handleAsteriskConnect;
                    _m_Manager.Disconnected -= handleAsteriskDisconnect;
                }

                _m_Manager = value;
                if (_m_Manager != null)
                {
                    _m_Manager.ManagerEvent += handleAsteriskEvent;
                    _m_Manager.Connected += handleAsteriskConnect;
                    _m_Manager.Disconnected += handleAsteriskDisconnect;
                }
            }
        }

        public AsteriskServer()
        {
            DMDObject.IncreaseCounter(this);
            m_Timer = new System.Timers.Timer();
            m_Timer.Interval = 60 * 5 * 1000;
            m_Timer.Enabled = false;
            m_Manager = null;
        }

        public AsteriskServer(string nome, string serverName, string channel, string userName, string password) : this()
        {
            Nome = nome;
            ServerName = serverName;
            Channel = channel;
            UserName = userName;
            Password = password;
        }

        public override string ToString()
        {
            return DMD.Strings.JoinW( Nome , " (" , ServerName , ":" , DMD.Strings.CStr(ServerPort), ")");
        }

        public AsteriskCallManager GetManager()
        {
            return m_Manager;
        }

        public void SetManager(AsteriskCallManager value)
        {
            m_Manager = value;
        }

        public bool IsConnected()
        {
            return m_Manager is object && GetManager().IsConnected();
        }

        public void Connect()
        {
            if (IsConnected())
                return;
            m_Manager = new AsteriskCallManager(UserName, Password, ServerName, ServerPort);
            m_Manager.Start();
            m_Manager.Login();
            m_Timer.Enabled = true;
        }

        public void Disconnect()
        {
            if (!IsConnected())
                return;
            m_Timer.Enabled = false;
            m_Manager.Stop();
            m_Manager = null;
        }

        private void handleAsteriskEvent(object sender, AsteriskEvent e)
        {
            ManagerEvent?.Invoke(sender, e);
        }

        private void handleAsteriskConnect(object sender, AsteriskEventArgs e)
        {
            Connected?.Invoke(sender, e);
        }

        private void handleAsteriskDisconnect(object sender, AsteriskEventArgs e)
        {
            Disconnected?.Invoke(sender, e);
        }

        private void m_Timer_Elapsed(object sender, System.Timers.ElapsedEventArgs e)
        {
            // If Me.IsConnected Then
            // Try

            // SyncLock Me
            // Me.Disconnect()
            // Me.Connect()
            // End SyncLock
            // Catch ex As Exception
            // minidom.Sistema.Events.NotifyUnhandledException(ex)
            // End Try
            // End If

        }

        public override bool Equals(object obj)
        {
            if (ReferenceEquals(obj, this))
                return true;
            if (!(obj is AsteriskServer))
                return false;
            {
                var withBlock = (AsteriskServer)obj;
                return (Nome ?? "") == (withBlock.Nome ?? "") && (ServerName ?? "") == (withBlock.ServerName ?? "") && ServerPort == withBlock.ServerPort && (UserName ?? "") == (withBlock.UserName ?? "") && (Password ?? "") == (withBlock.Password ?? "") && (Channel ?? "") == (withBlock.Channel ?? "") && (CallerID ?? "") == (withBlock.CallerID ?? "");
            }
        }

        protected virtual void SetFieldInternal(string fieldName, object fieldValue)
        {
            switch (fieldName ?? "")
            {
                case "nome":
                    {
                        Nome = DMD.XML.Utils.Serializer.DeserializeString(fieldValue);
                        break;
                    }

                case "servername":
                    {
                        ServerName = DMD.XML.Utils.Serializer.DeserializeString(fieldValue);
                        break;
                    }

                case "serverport":
                    {
                        ServerPort = (int)DMD.XML.Utils.Serializer.DeserializeInteger(fieldValue);
                        break;
                    }

                case "username":
                    {
                        UserName = DMD.XML.Utils.Serializer.DeserializeString(fieldValue);
                        break;
                    }

                case "password":
                    {
                        Password = DMD.XML.Utils.Serializer.DeserializeString(fieldValue);
                        break;
                    }

                case "channel":
                    {
                        Channel = DMD.XML.Utils.Serializer.DeserializeString(fieldValue);
                        break;
                    }

                case "callerid":
                    {
                        CallerID = DMD.XML.Utils.Serializer.DeserializeString(fieldValue);
                        break;
                    }
            }
        }

        protected virtual void XMLSerialize(XMLWriter writer)
        {
            writer.WriteAttribute("nome", Nome);
            writer.WriteAttribute("servername", ServerName);
            writer.WriteAttribute("serverport", ServerPort);
            writer.WriteAttribute("username", UserName);
            writer.WriteAttribute("password", Password);
            writer.WriteAttribute("channel", Channel);
            writer.WriteAttribute("callerid", CallerID);
        }

        ~AsteriskServer()
        {
            DMDObject.DecreaseCounter(this);
        }

        // Private Sub ReadXml(reader As Xml.XmlReader) Implements Xml.Serialization.IXmlSerializable.ReadXml
        // Me.Nome = reader.GetAttribute("nome")
        // Me.ServerName = reader.GetAttribute("servername")
        // Me.ServerPort = CInt(reader.GetAttribute("serverport"))
        // Me.UserName = reader.GetAttribute("username")
        // Me.Password = reader.GetAttribute("password")
        // Me.Channel = reader.GetAttribute("channel")
        // Me.CallerID = reader.GetAttribute("callerid")
        // reader.Read()
        // End Sub

        // Public Sub WriteXml(writer As Xml.XmlWriter) Implements Xml.Serialization.IXmlSerializable.WriteXml
        // writer.WriteAttributeString("nome", Me.Nome)
        // writer.WriteAttributeString("servername", Me.ServerName)
        // writer.WriteAttributeString("serverport", CStr(Me.ServerPort))
        // writer.WriteAttributeString("username", Me.UserName)
        // writer.WriteAttributeString("password", Me.Password)
        // writer.WriteAttributeString("channel", Me.Channel)
        // writer.WriteAttributeString("callerid", Me.CallerID)
        // End Sub
    }


}

File: MiniDOMSIP/Config/AsteriskServer.vb
Imports minidom
Imports minidom.Sistema
Imports minidom.CallManagers
Imports minidom.CallManagers.Events

Public Class AsteriskServer
    Implements XML.IDMDXMLSerializable

    Public Event Connected(ByVal sender As Object, ByVal e As AsteriskEventArgs)
    Public Event Disconnected(ByVal sender As Object, ByVal e As AsteriskEventArgs)
    Public Event ManagerEvent(ByVal sender As Object, ByVal e As AsteriskEvent)


    Private WithEvents m_Timer As System.Timers.Timer

    Public Nome As String
    Public ServerName As String
    Public ServerPort As Integer = 5038
    Public UserName As String = ""
    Public Password As String = ""
    Public Channel As String = ""
    Public CallerID As String = ""
    Private WithEvents m_Manager As minidom.CallManagers.AsteriskCallManager

    Public Sub New()
        DMDObject.IncreaseCounter(Me)
        Me.m_Timer = New System.Timers.Timer
        Me.m_Timer.Interval = 60 * 5 * 1000
        Me.m_Timer.Enabled = False
        Me.m_Manager = Nothing
    End Sub

    Public Sub New(ByVal nome As String, ByVal serverName As String, ByVal channel As String, ByVal userName As String, ByVal password As String)
        Me.New()
        Me.Nome = nome
        Me.ServerName = serverName
        Me.Channel = channel
        Me.UserName = userName
        Me.Password = password
    End Sub

    Public Overrides Function ToString() As String
        Return Me.Nome & " (" & Me.ServerName & ":" & Me.ServerPort & ")"
    End Function



    Public Function GetManager() As minidom.CallManagers.AsteriskCallManager
        Return Me.m_Manager
    End Function

    Public Sub SetManager(ByVal value As minidom.CallManagers.AsteriskCallManager)
        Me.m_Manager = value
    End Sub

    Public Function IsConnected() As Boolean
        Return Me.m_Manager IsNot Nothing AndAlso Me.GetManager.IsConnected
    End Function

    Public Sub Connect()
        If Me.IsConnected Then Exit Sub
        Me.m_Manager = New minidom.CallManagers.AsteriskCallManager(Me.UserName, Me.Password, Me.ServerName, Me.ServerPort)

        Me.m_Manager.Start()
        Me.m_Manager.Login()

        Me.m_Timer.Enabled = True
    End Sub

    Public Sub Disconnect()
        If Not Me.IsConnected Then Return
        Me.m_Timer.Enabled = False

        Me.m_Manager.Stop()
        Me.m_Manager = Nothing

    End Sub


    Private Sub handleAsteriskEvent(sender As Object, e As AsteriskEvent) Handles m_Manager.ManagerEvent
        RaiseEvent ManagerEvent(sender, e)
    End Sub


    Private Sub handleAsteriskConnect(sender As Object, e As AsteriskEventArgs) Handles m_Manager.Connected
        RaiseEvent Connected(sender, e)
    End Sub

    Private Sub handleAsteriskDisconnect(sender As Object, e As AsteriskEventArgs) Handles m_Manager.Disconnected
        RaiseEvent Disconnected(sender, e)
    End Sub

    Private Sub m_Timer_Elapsed(sender As Object, e As Timers.ElapsedEventArgs) Handles m_Timer.Elapsed
        'If Me.IsConnected Then
        '    Try

        '        SyncLock Me
        '            Me.Disconnect()
        '            Me.Connect()
        '        End SyncLock
        '    Catch ex As Exception
        '        minidom.Sistema.Events.NotifyUnhandledException(ex)
        '    End Try
        'End If

    End Sub

    Public Overrides Function Equals(obj As Object) As Boolean
        If (obj Is Me) Then Return True
        If Not (TypeOf (obj) Is AsteriskServer) Then Return False
        With DirectCast(obj, AsteriskServer)
            Return Me.Nome = .Nome AndAlso
                   Me.ServerName = .ServerName AndAlso
                   Me.ServerPort = .ServerPort AndAlso
                   Me.UserName = .UserName AndAlso
                   Me.Password = .Password AndAlso
                   Me.Channel = .Channel AndAlso
                   Me.CallerID = .CallerID
        End With
    End Function

    Protected Overridable Sub SetFieldInternal(fieldName As String, fieldValue As Object) Implements XML.IDMDXMLSerializable.SetFieldInternal
        Select Case fieldName
            Case "nome" : Me.Nome = XML.Utils.Serializer.DeserializeString(fieldValue)
            Case "servername" : Me.ServerName = XML.Utils.Serializer.DeserializeString(fieldValue)
            Case "serverport" : Me.ServerPort = CInt(XML.Utils.Serializer.DeserializeInteger(fieldValue))
            Case "username" : Me.UserName = XML.Utils.Serializer.DeserializeString(fieldValue)
            Case "password" : Me.Password = XML.Utils.Serializer.DeserializeString(fieldValue)
            Case "channel" : Me.Channel = XML.Utils.Serializer.DeserializeString(fieldValue)
            Case "callerid" : Me.CallerID = XML.Utils.Serializer.DeserializeString(fieldValue)
        End Select
    End Sub

    Protected Overridable Sub XMLSerialize(writer As minidom.XML.XMLWriter) Implements XML.IDMDXMLSerializable.XMLSerialize
        writer.WriteAttribute("nome", Me.Nome)
        writer.WriteAttribute("servername", Me.ServerName)
        writer.WriteAttribute("serverport", Me.ServerPort)
        writer.WriteAttribute("username", Me.UserName)
        writer.WriteAttribute("password", Me.Password)
        writer.WriteAttribute("channel", Me.Channel)
        writer.WriteAttribute("callerid", Me.CallerID)
    End Sub

    Protected Overrides Sub Finalize()
        MyBase.Finalize()
        DMDObject.DecreaseCounter(Me)
    End Sub

    'Private Sub ReadXml(reader As Xml.XmlReader) Implements Xml.Serialization.IXmlSerializable.ReadXml
    '    Me.Nome = reader.GetAttribute("nome")
    '    Me.ServerName = reader.GetAttribute("servername")
    '    Me.ServerPort = CInt(reader.GetAttribute("serverport"))
    '    Me.UserName = reader.GetAttribute("username")
    '    Me.Password = reader.GetAttribute("password")
    '    Me.Channel = reader.GetAttribute("channel")
    '    Me.CallerID = reader.GetAttribute("callerid")
    '    reader.Read()
    'End Sub

    'Public Sub WriteXml(writer As Xml.XmlWriter) Implements Xml.Serialization.IXmlSerializable.WriteXml
    '    writer.WriteAttributeString("nome", Me.Nome)
    '    writer.WriteAttributeString("servername", Me.ServerName)
    '    writer.WriteAttributeString("serverport", CStr(Me.ServerPort))
    '    writer.WriteAttributeString("username", Me.UserName)
    '    writer.WriteAttributeString("password", Me.Password)
    '    writer.WriteAttributeString("channel", Me.Channel)
    '    writer.WriteAttributeString("callerid", Me.CallerID)
    'End Sub
End Class
